Avatar implementation

YeahImRose opened the pull request:
A squashed version of PR #92

YeahImRose committed:
Squashed avatar implementation

YeahImRose committed:
Squashed avatar implementation

peppy commented on osu.Game/Graphics/Avatar.cs:
```diff
@@ -0,0 +1,65 @@
+    public class Avatar : Sprite
+    {
+        private int userId;
+        private int avatarSize = 50;
```
this is silly. why not just use .Size or .Scale?

peppy commented on osu.Game/Overlays/Toolbar.cs:
```diff
@@ -87,26 +121,36 @@ public override void Load(BaseGame game)
+
+            InitUser();
+            Scheduler.AddDelayed(delegate {
+                InitUser();
```
what?

peppy commented on osu.Game/Overlays/Toolbar.cs:
```diff
@@ -39,13 +47,39 @@ protected override void PopOut()
 
-            RelativeSizeAxes = Axes.X;
-            Size = new Vector2(1, height);
+        private void InitUser()
```
you're calling this every 10 seconds with no checks for whether it's still running, failure of last call, etc.

what?
what?
what?

peppy commented on osu.Game/Overlays/Toolbar.cs:
```diff
@@ -39,13 +47,39 @@ protected override void PopOut()
+            api.Queue(req);
+            req.Success += delegate (User MyUser)
+            {
+                this.user = MyUser;
```
you're throwing out the old user and replacing it every time this is called. please no.

peppy commented on osu.Game/Graphics/Avatar.cs:
```diff
@@ -0,0 +1,65 @@
+            {
+                try
+                {
+                    using (var wc = new System.Net.WebClient())
```
why are you blatantly ignoring the code i wrote for async retrieval, which also has error handling, timeout checking and protocol fallback support?

also there's no debounce

peppy commented on osu.Game/Graphics/Avatar.cs:
```diff
@@ -0,0 +1,65 @@
+            {
+                if (imageData != null)
+                {
+                    Texture = TextureLoader.FromBytes(imageData);
```
not using atlases

peppy commented on osu.Game/Graphics/Avatar.cs:
```diff
@@ -0,0 +1,65 @@
+
+        public override void Load(BaseGame game)
+        {
+            this.game = game;
```
why? you never use it?

peppy commented on osu.Game/Overlays/Toolbar.cs:
```diff
@@ -87,26 +121,36 @@ public override void Load(BaseGame game)
+        protected override void Update()
+        {
+            base.Update();
+            api.Update();
```
this is not intended use. you're taking Game's api and then updating it here a second time.

peppy commented on osu.Game/Overlays/Toolbar.cs:
```diff
@@ -87,26 +121,36 @@ public override void Load(BaseGame game)
+            InitUser();
+            Scheduler.AddDelayed(delegate {
+                InitUser();
+            }, 10000, true);
```
scheduling with no cancel logic?

peppy commented on osu.Game/Overlays/Toolbar.cs:
```diff
@@ -39,13 +47,39 @@ protected override void PopOut()
+        {
+            base.Load(game);
+            api = ((OsuGameBase)game).API;
+            this.game = game;
```
not using this, don't store.

peppy commented:
you don't need to add "squashed" in commit messages.
also you should work on explaining what you are doing in the PR better.
also you know you can rebase/force push rather than opening a new PR, right?

Lunar-Dust commented:
I had over 130 commits that I'd have to manually add due to merge conflicts. I know rebasing is a thing. That's what I tried, and it was literally going to take hours. This was a much faster solution. I'll do some more work on this and update it later this week probably.

peppy commented:
i meant more that you could force push your new head on your old PR, not that you interactive-rebase one step at a time.

YeahImRose committed:
Async and rework

YeahImRose committed:
Async and rework

YeahImRose committed:
Update osu.Game.csproj

YeahImRose committed:
Async and rework

I know this probably still isn't perfect, but I'd like some feedback on what could be done better :)

YeahImRose committed:
Remove unneeded changes

YeahImRose committed:
Update to latest code version

YeahImRose committed:
Merge branch 'master' into settings-wip

Fix merge conflicts

YeahImRose committed:
Update csproj file

YeahImRose committed:
More merge

YeahImRose committed:
Update everything to latest code version

YeahImRose committed:
Cleanup all of osu.Game.csproj

YeahImRose committed:
Cleanup all of osu.Game.csproj

Alphabetize and stuff

peppy commented:
please fix your commits and come back if/when done.

Lunar-Dust commented:
Re-did my commits and made intelligent commit messages, down to 6 commits now
